bot/seasons/evergreen/bookmark.py

```diff
@@ -57,6 +61,28 @@ def __init__(self, bot: commands.Bot):
         else:
             log.info(f"{ctx.author} bookmarked {target_message.jump_url} with title '{title}'")
             await ctx.message.add_reaction(Emojis.envelope)
+            await ctx.message.add_reaction(Emojis.pin)
+
+        embed.add_field(
+            name=f'Bookmarked from {ctx.author.name}.',
+            value=f'[Visit original message]({ctx.message.jump_url})',
+            inline=False
+        )
+
+        def check(reaction: discord.Reaction, user: discord.User) -> bool:
+            return reaction.emoji == Emojis.pin
+
+        sent_person = [ctx.author]  # list of id who got the message
+        while True:
+            try:
+                reaction, user = await self.bot.wait_for("reaction_add", timeout=60.0, check=check)
+                if user != self.bot.user and user not in sent_person:

```

Comment:
Since `Member` and `User` are both hashable, you can safely use a set here, as it will tremendously improve the membership check.

With that logic in mind, you can move all of them together into the `check` since it is also part of checking who reacted, like so:
```py
def check(reaction: discord.Reaction, user: discord.User) -> bool:
    return (
        user != self.bot.user and
        reaction.emoji == Emojis.pin and
        user not in sent_person
    )

sent_person = {ctx.author}
```

Then your loop will not have any extra indentation, simply
```py
while True:
    try:
        reaction, user = await self.bot.wait_for("reaction_add", timeout=60.0, check=check)
        log.info(f"{user.name} bookmarked {target_message.jump_url} with title '{title}' from {ctx.author}")
        await user.send(embed=embed)
        sent_person.append(user)
    except asyncio.TimeoutError:
        ....
```

Reply:
Also, there should also be a check where the message with the reaction is the same message as `ctx.message` - Otherwise the bot will simply check for reaction everywhere.

Reply:
Please don't auto resolve, you can certainly ping me in the discord channel for discussion / a review over it as well.

As it stands, you can certainly change `sent_person` from being a list to a set since `Member` and `User` are hashable, do you have plans for that as well?

Reply:
no , list are way better and make it easy to read , so I am gonna keep that as list only.

Reply:
You are not reading it anywhere at all, you only keep it to do a membership checking - avoid sending to the same person twice, is that correct?

Reply:
We are reading it in the check ,also i feel like someone in future ever read this code , he is gonna get confused with sets.

Reply:
If that is a case, I strongly recommend going from a list to a set, as set has a constant, close to O(1) for membership checking as opposed to a list with a O(n) to do the same thing - unless we are doing something else with the list.

Reply:
I doubt anyone will get confused by sets; they read just as easy as lists and make absolute sense in this context.

Reply:
Commited